### PR TITLE
Add storage metrics that count records that failed to insert

### PR DESCRIPTION
**Pull request by rtm0**
### Describe Your Changes

Add storage metrics that count records that failed to insert:

-   `RowsReceivedTotal`: the number of records that have been received by the storage from the clients
-   `RowsAddedTotal`: the number of records that have actually been persisted. This value must be equal to `RowsReceivedTotal` if all the records have been	valid ones. But it will be smaller otherwise. The values of the metrics below should provide the insight of why some records hasn't been added
-   `NaNValueRows`: the number of records whose value was `NaN`
-   `StaleNaNValueRows`: the number of records whose value was `Stale NaN`
-   `InvalidRawMetricNames`: the number of records whose raw metric name has failed to unmarshal.

The following metrics existed before this PR and are listed here for completeness:

-   `TooSmallTimestampRows`: the number of records whose timestamp is negative or is older than retention period
-   `TooBigTimestampRows`: the number of records whose timestamp is too far in the future.
-   `HourlySeriesLimitRowsDropped`: the number of records that have not been 	added because the hourly series limit has been exceeded.
-   `DailySeriesLimitRowsDropped`: the number of records that have not been added because the daily series limit has been exceeded.

### Checklist

The following checks are **mandatory**:

- [x ] My change adheres [VictoriaMetrics contributing guidelines](https://docs.victoriametrics.com/contributing/).

### CODE REVIEW

**Comment by rtm0** [comment]
A friendly ping.

File: app/vmstorage/main.go [hunk 1]
```diff
@@ -521,12 +521,22 @@ func writeStorageMetrics(w io.Writer, strg *storage.Storage) {
 	metrics.WriteGaugeUint64(w, `vm_data_size_bytes{type="indexdb/inmemory"}`, idbm.InmemorySizeBytes)
 	metrics.WriteGaugeUint64(w, `vm_data_size_bytes{type="indexdb/file"}`, idbm.FileSizeBytes)
 
+	metrics.WriteCounterUint64(w, `vm_rows_received_by_storage_total`, m.RowsReceivedTotal)
```
Comment (hagen1778) [inline app/vmstorage/main.go:524]: Seems like this metric is already represented in form of `vm_vminsert_metrics_read_total`

**Review by hagen1778 — COMMENTED** [review]
Thanks for the PR!
@zekker6 @f41gh7 I'd appreciate your review

File: app/vmstorage/main.go (hunk 1 above)
Comment (hagen1778) [reply app/vmstorage/main.go:524]: This metic is defined in [cluster branch](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/91d989d2d16195f42095bd834f884b0766524151/app/vmstorage/servers/vminsert.go#L143).
In VM we have two branches: master and cluster. Where master is for single-node and cluster is for clustered version. They share about 95% of the source code, but still contain some cluster-related differences.

Comment (rtm0) [reply app/vmstorage/main.go:524]: I see. But it feels like `vm_vminsert_metrics_read_total` count represents the number of records sent to the cluster as a whole. While the new metric will count the number of received metrics per replica. I.e. a single replica would have all the counts in one place so the dropped records could be diagnosed more conveniently.

File: app/vmstorage/main.go [hunk 2]
```diff
@@ -521,12 +521,22 @@ func writeStorageMetrics(w io.Writer, strg *storage.Storage) {
...
 	metrics.WriteCounterUint64(w, `vm_rows_added_to_storage_total`, m.RowsAddedTotal)
 	metrics.WriteCounterUint64(w, `vm_deduplicated_samples_total{type="merge"}`, m.DedupsDuringMerge)
 	metrics.WriteGaugeUint64(w, `vm_snapshots`, m.SnapshotsCount)
 
+	metrics.WriteCounterUint64(w, `vm_rows_ignored_total{reason="nan_value"}`, m.NaNValueRows)
+	metrics.WriteCounterUint64(w, `vm_rows_ignored_total{reason="stale_nan_value"}`, m.StaleNaNValueRows)
 	metrics.WriteCounterUint64(w, `vm_rows_ignored_total{reason="big_timestamp"}`, m.TooBigTimestampRows)
 	metrics.WriteCounterUint64(w, `vm_rows_ignored_total{reason="small_timestamp"}`, m.TooSmallTimestampRows)
+	metrics.WriteCounterUint64(w, `vm_rows_ignored_total{reason="invalid_raw_metric_name"}`, m.InvalidRawMetricNames)
```
Comment (f41gh7) [inline app/vmstorage/main.go:533]: Looks like this metric is not really useful. It may indicate about internal bug of VictoriaMetrics components.

 I think, logging is more then enough to cover this case.

**Review by f41gh7 — COMMENTED** [review]
VictoriaMetrics already exposes a lot of different metrics. And I suggest to add new metrics only with reasonable purpose.

 Such as to me:
- troubleshooting case of unexpected behavior of components. 
- actionable metric for alerting rules.
- behavior visualization metrics.


 It would be great if you could add examples of usage for added metrics.

**Comment by rtm0** [comment]
> I suggest to add new metrics only with reasonable purpose.

Since VictoriaMetrics is a database, one of its primary purposes is storing the data, I believe that these new metrics fit all the purposes that you have enumerated:

>     * troubleshooting case of unexpected behavior of components.

If query shows incorrect results, then one of the first things to check is whether the database have inserted records. If the rate of dropped records is high, then chances are that the data hasn't been written. 

>     * actionable metric for alerting rules.

I understand that VictoriaMetrics does not make any durability guarantees (i.e. that the accepted data has been persisted) but it might be useful to know just by how much the number of inserted records differs from the number of the records sent to the database. Depending on what durability expectations a project has, one could set up alerts on smaller or higher percent of dropped records and then investigate the exact reason. 

>     * behavior visualization metrics.

This could be something as simple as the ratio of `rows_added_total / rows_received_total` and as complex as a stacked diagram showing all the reasons of dropped records. For example: 
stacked


Ideally, the number of received records must be equal to the number of added records minus number of dropped records. However, the AddRows() and RegisterMetricNames() share the same metrics to count reasons for dropped records. So if RegisterMetricNames() is in use, then the equation will not hold. This can be addressed by introducing a separate set of metrics for RegisterMetricNames().

File: app/vmstorage/main.go [hunk 3]
```diff
@@ -521,12 +521,22 @@ func writeStorageMetrics(w io.Writer, strg *storage.Storage) {
 	metrics.WriteGaugeUint64(w, `vm_data_size_bytes{type="indexdb/inmemory"}`, idbm.InmemorySizeBytes)
 	metrics.WriteGaugeUint64(w, `vm_data_size_bytes{type="indexdb/file"}`, idbm.FileSizeBytes)
 
+	metrics.WriteCounterUint64(w, `vm_rows_received_by_storage_total`, m.RowsReceivedTotal)
 	metrics.WriteCounterUint64(w, `vm_rows_added_to_storage_total`, m.RowsAddedTotal)
 	metrics.WriteCounterUint64(w, `vm_deduplicated_samples_total{type="merge"}`, m.DedupsDuringMerge)
 	metrics.WriteGaugeUint64(w, `vm_snapshots`, m.SnapshotsCount)
 
+	metrics.WriteCounterUint64(w, `vm_rows_ignored_total{reason="nan_value"}`, m.NaNValueRows)
```
[the remark this answers is not in the document]
Comment (rtm0) [reply app/vmstorage/main.go:529]: I agree that under normal conditions these two are unlikely to happen. And it is hard to come up with the case when a significant number of records is dropped because of this. A bug in the client code that sends NaN values only? A bug in VM that treats all values as NaNs? Both are very unlikely.

Yet the VM code conciders this cases and therefore they should be made visible somehow. Logging may not be enough because not all messages are being logged as I've been told.

File: app/vmstorage/main.go [hunk 4]
```diff
@@ -521,12 +521,22 @@ func writeStorageMetrics(w io.Writer, strg *storage.Storage) {
...
+	metrics.WriteCounterUint64(w, `vm_rows_ignored_total{reason="stale_nan_value"}`, m.StaleNaNValueRows)
 	metrics.WriteCounterUint64(w, `vm_rows_ignored_total{reason="big_timestamp"}`, m.TooBigTimestampRows)
 	metrics.WriteCounterUint64(w, `vm_rows_ignored_total{reason="small_timestamp"}`, m.TooSmallTimestampRows)
+	metrics.WriteCounterUint64(w, `vm_rows_ignored_total{reason="invalid_raw_metric_name"}`, m.InvalidRawMetricNames)
+	if *maxHourlySeries > 0 {
+		metrics.WriteCounterUint64(w, `vm_rows_ignored_total{reason="hourly_limit_exceeded"}`, m.HourlySeriesLimitRowsDropped)
```
[the remark this answers is not in the document]
Comment (rtm0) [reply app/vmstorage/main.go:535]: They are exposed indeed. But they fit the `ignored rows` metric better. Do you think we can remove the existing ones or it might result in backward compatibility issues?

File: app/vmstorage/main.go (hunk 2 above)
Comment (rtm0) [reply app/vmstorage/main.go:533]: Same here. See my comment on NaNs.

File: app/vmstorage/main.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (rtm0) [reply app/vmstorage/main.go:529]: To put it more simply these would be very useful for debugging.

File: app/vmstorage/main.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (rtm0) [reply app/vmstorage/main.go:535]: Ok, it looks like we can't delete the existing metrics because someone could be using them. But having these two included as reasons into the `vm_rows_ignored_total` metric could significantly simplify troubleshooting all the reasons why samples ignored. I.e. the queries will be much simpler, for example:

```
sum(rate(vm_rows_ignored_total)) by (reason)
```

**Review by hagen1778 — COMMENTED** [review]
I think most of these new metrics have sense for debugability purposes.
But we need to make sure they're indeed actionable for users.
That being said, we should verify that alert `increse(vm_rows_ignored_total) > 0` won't trigger for the new release with the new added labels. VM and VM cluster need to be tested with this code change to see if `nan`s or `stale nan`s are expected or not.

**Comment by f41gh7** [comment]
Also, this PR should cover case, when multiple insert clusters connected to the single vmstorage cluster. In this case, it could be tricky to collect metrics only from vminsert.

 I also think, we should add additional metrics related to the vmselect requests. First of all summary for metricID/tsid per search request.

**Comment by rtm0** [comment]
> I think most of these new metrics have sense for debugability purposes. But we need to make sure they're indeed actionable for users. That being said, we should verify that alert `increse(vm_rows_ignored_total) > 0` won't trigger for the new release with the new added labels. VM and VM cluster need to be tested with this code change to see if `nan`s or `stale nan`s are expected or not.

I've tested it locally (docker-compose with single node). I was sending NaNs to the VM instance and this indeed triggered an alert. @hagen1778 and I agreed that we should filter NaNs at the vminsert side and introduce a separate vminsert metric for that. This will be a separate PR that should go in before this one. We also agreed that the other two (staleNaNs and metricNameRaw unmarshalling errors) should still be handled by the vmstorage. Both mean a broken client and broken VM code respectively.

Will link the PR here once it is ready.

**Comment by rtm0** [comment]
Ok, per another discussion with @hagen1778 and @valyala we have decided to keep the NaN check as is and do not drop them in vminsert. The assumptions are the following:

1.  Clients normally don't send NaNs and the [RowsRejectedOnIngestion](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/master/deployment/docker/alerts.yml#L60) alert is not expected to fire. If we see otherwise, we will come up with a solution.
2.  Values need to be checked at one place (vmstorage). Checking values in vminsert and vmstorage leads to the code and metrics that hard to follow. vminsert does perform checks as well, but those thecks are related to parsing the incoming data streams.

What needs to be removed however is the stale_nan reason because the code that increments that counter is being removed in #6936. Thus, waiting until that PR is merged.

**Comment by rtm0** [comment]
@f41gh7 / @hagen1778 I have removed stale nans. At this point the PR is ready to be merged. Please take a look.

**Comment by hagen1778** [comment]
@valyala @rtm0 metric `NaNValueRows: the number of records whose value was NaN` was rolled back in
https://github.com/VictoriaMetrics/VictoriaMetrics/commit/0d4f4b8f7d1ce979d6c227a7b823c0395a6575d5. See commit description for reasoning.
